CMakeLists.txt

```
@@ -0,0 +1,129 @@
+cmake_minimum_required(VERSION 3.23)
+
+project(llm.c
+    LANGUAGES C
+    VERSION 0.1.0
+)
+
+set(CMAKE_C_STANDARD 11)
+set(CMAKE_C_STANDARD_REQUIRED True)
+
+set(GPT2_TRAIN_SOURCES platform_utils.h)
+set(GPT2_TRAIN_DEFINES)
+set(GPT2_TEST_SOURCES platform_utils.h)
+set(GPT2_LIB_DEPENDENCIES)
+set(GPT2_CUSTOM_DEPENDENCIES)
+set(GPT2_COMPILE_FLAGS)
+
+# optional cuda support
+if(ENABLE_CUDA)
+    find_package(CUDAToolkit)
+    if(NOT CUDAToolkit_FOUND AND WIN32)
+        message(SEND_ERROR "CUDA is not installed. Go to https://developer.nvidia.com/cuda-toolkit")
+    endif()
+
+    enable_language(CUDA)
+    set(CMAKE_CUDA_STANDARD 11)
+    set(CMAKE_CUDA_STANDARD_REQUIRED True)
+
+    list(APPEND GPT2_TRAIN_SOURCES train_gpt2.cu)
+    list(APPEND GPT2_TEST_SOURCES test_gpt2.cu)
+
+    list(APPEND GPT2_LIB_DEPENDENCIES CUDA::cublas)
+
+    # https://forums.developer.nvidia.com/t/suppress-warning-conversion-from-a-string-literal-to-char-is-deprecated/258050/3
+    list(APPEND GPT2_COMPILE_FLAGS "-diag-suppress" "2464")
+else()
+    list(APPEND GPT2_TRAIN_SOURCES train_gpt2.c)
+    list(APPEND GPT2_TEST_SOURCES test_gpt2.c)
+
+    find_package(OpenMP)
```

```suggestion
    find_package(OpenMP 5.2)
```

Avoid involving the un-appropriate openmp building at the very beginning